lib/compress/zstd_opt.c

```diff
@@ -764,6 +764,151 @@ FORCE_INLINE_TEMPLATE U32 ZSTD_BtGetAllMatches (
+    literalsBytesRemaining = (optLdm->seqStore.posInSequence < currSeq.litLength) ?
+            currSeq.litLength - (U32)optLdm->seqStore.posInSequence :
+            0;
+    matchBytesRemaining = (literalsBytesRemaining == 0) ?
+            currSeq.matchLength - ((U32)optLdm->seqStore.posInSequence - currSeq.litLength) :
+            currSeq.matchLength;
+
+    /* If there are more literal bytes than bytes remaining in block, no ldm is possible */
+    if (literalsBytesRemaining >= blockBytesRemaining) {
+        optLdm->startPosInBlock = UINT_MAX;
+        optLdm->endPosInBlock = UINT_MAX;
+        ZSTD_optLdm_skipRawSeqStoreBytes(&optLdm->seqStore, blockBytesRemaining);
+        return;
+    }
+
+    /* Matches may be < MINMATCH by this process. In that case, we will reject them
+       when we are deciding whether or not to add the ldm */
+    optLdm->startPosInBlock = currPosInBlock + literalsBytesRemaining;
+    optLdm->endPosInBlock = optLdm->startPosInBlock + matchBytesRemaining;
+    optLdm->offset = currSeq.offset;
+
+    if (optLdm->endPosInBlock > currBlockEndPos) {
+        /* Match ends after the block ends, we can't use the whole match */
+        optLdm->endPosInBlock = currBlockEndPos;
+        ZSTD_optLdm_skipRawSeqStoreBytes(&optLdm->seqStore, currBlockEndPos - currPosInBlock);
+    } else {
+        /* Consume nb of bytes equal to size of sequence left */
+        ZSTD_optLdm_skipRawSeqStoreBytes(&optLdm->seqStore, literalsBytesRemaining + matchBytesRemaining);
+    }
+}
+
+/* ZSTD_optLdm_maybeAddMatch():
+ * Adds a match if it's long enough, based on it's 'matchStartPosInBlock'
+ * and 'matchEndPosInBlock', into 'matches'. Maintains the correct ordering of 'matches'
+ */
+static void ZSTD_optLdm_maybeAddMatch(ZSTD_match_t* matches, U32* nbMatches,
+                                      ZSTD_optLdm_t* optLdm, U32 currPosInBlock) {
+    U32 posDiff = currPosInBlock - optLdm->startPosInBlock;
+    /* Note: ZSTD_match_t actually contains offCode and matchLength (before subtracting MINMATCH) */
+    U32 candidateMatchLength = optLdm->endPosInBlock - optLdm->startPosInBlock - posDiff;
+    U32 candidateOffCode = optLdm->offset + ZSTD_REP_MOVE;
+
+    /* Ensure that current block position is not outside of the match */
+    if (currPosInBlock < optLdm->startPosInBlock
+      || currPosInBlock >= optLdm->endPosInBlock
+      || candidateMatchLength < MINMATCH) {
+        return;
+    }
+
+    DEBUGLOG(6, "ZSTD_optLdm_maybeAddMatch(): Adding ldm candidate match (offCode: %u matchLength %u) at block position=%u",
+             candidateOffCode, candidateMatchLength, currPosInBlock);
+    if (*nbMatches == 0) {
+        matches[*nbMatches].len = candidateMatchLength;
+        matches[*nbMatches].off = candidateOffCode;
+        (*nbMatches)++;
+    } else if ((candidateMatchLength >= matches[*nbMatches-1].len) && *nbMatches < ZSTD_OPT_NUM) {
+        if (candidateMatchLength == matches[*nbMatches-1].len) {
```

Comment:
This doesn't seem worthwhile.
This case seems very rare, and offer limited, if any, benefit.

I guess it means this could be simplified : 
only take care of the case when `candidateMatchLength` is strictly `>` longer.

This is a very minor tweak though : I don't expect either runtime nor compression to be significantly impacted.

Reply:
Yeah, I measured a few hundred bytes saved on `silesia.tar` which is why I kept it in since it seemed like a more or less "free" gain, but I'm fine removing this check if making the code simpler is worth that.

Reply:
In my tests, the gain from this section of the code (`==` matchlength) is variable, sometimes better, sometimes worse.
For `silesia.tar`, at `-19 --long=23`, strict `>` is smaller by 3 bytes.
But indeed, at `--ultra -22 --long`, strict `>` is larger by ~150 bytes.

If you want to keep it, consider a variant which tests : `if (4*candidateOffCode < matches[*nbMatches-1].off)`

i.e. only replace the existing match by the new one if it is _noticeably_ cheaper to encode.

I tested this variant, and it seems more consistently positive,
including on `silesia.tar`, at all compression settings tested.
At `--ultra -22 --long`, it saves an additional KB.

Reply:
Good idea! It's interesting that the long distance match finder in certain cases can even produce matches with offsets multiple times smaller than what the regular match finder produces - it's definitely a little surprising..

Reply:
I just measured myself, and oddly couldn't replicate the same results on `silesia.tar`:

With `-22 --ultra --long`:
`candidateOffCode < matches[*nbMatches-1].off <: 52694897`
`2*candidateOffCode < matches[*nbMatches-1].off <: 52695398`
`4*candidateOffCode < matches[*nbMatches-1].off <: 52695296`
`candidateMatchLength > matches[*nbMatches-1].len : 52695282`

With `-19 --long`:
`candidateOffCode < matches[*nbMatches-1].off < : 53032567`
`2*candidateOffCode < matches[*nbMatches-1].off <: 53034038`
`4*candidateOffCode < matches[*nbMatches-1].off < : 53033516`
`candidateMatchLength > matches[*nbMatches-1].len : 53032764`

On concatenated linux kernels with `-22 --ultra --long=30`
`candidateOffCode < matches[*nbMatches-1].off < : 120561718`
`2*candidateOffCode < matches[*nbMatches-1].off <: 120563852`
`4*candidateOffCode < matches[*nbMatches-1].off <: 120555259`
`candidateMatchLength > matches[*nbMatches-1].len : 120564642`

Despite the `silesia.tar` results, it seems like on the case that is most practical (large, similar files concatenated), the `4*candidateOffCode` change is most beneficial by a fair bit.

Reply:
Indeed, we don't experience the same results.

I presume that's because `silesia.tar` actually doesn't exist,
the `silesia` corpus is distributed as one or several `.zip` files,
and the `.tar` archive must be rebuild locally, using whatever local version of `tar`.
Which means that 2 different systems can produce slightly different `tar` archives.
It's enough to produce measurable differences.

It wouldn't matter as much if the overall direction was always the same,
but on such tiny changes, it ends up measuring benefits on one side, and losses on the other one,
so it's confusing.

I also agree that, since results remain so close to each other, the simplified "strict `>`" version feels more appealing
(though I'm surprised that difference can be as much as 10KB for the concatenated linux kernel).